File: leetcode/roman_to_integer.js
/* Given a roman numeral, convert it to an integer.
 * Input is guaranteed to be within the range from 1 to 3999.
 * Time complexity: O(n)
 * Space complexity: O(1)
 */

/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    var map = new Map();
    map.set('M', 1000);
    map.set('D', 500);
    map.set('C', 100);
    map.set('L', 50);
    map.set('X', 10);
    map.set('V', 5);
    map.set('I', 1);
    var result = map.get(s.charAt(s.length - 1));
    var previous = result;
    for (var i = s.length - 2; i >= 0; --i) {
        var val = map.get(s.charAt(i));
        result += val >= previous ? val : -val;
        previous = val;
    }
    return result;
};

File: lintcode/roman_to_integer.py
class Solution:
    # @param {string} s Roman representation
    # @return {int} an integer
    def romanToInt(self, s):
        map = [0] * 256
        map[ord('M')], map[ord('D')], map[ord('C')] = 1000, 500, 100
        map[ord('L')], map[ord('X')], map[ord('V')] = 50, 10, 5
        map[ord('I')] = 1
        previous = result = map[ord(s[-1])]
        for index in xrange(len(s) - 2, -1, -1):
            val = map[ord(s[index])]
            result += val if val >= previous else -val
            previous = val
        return result
